File: python/zeroipc/latch.py
"""Lock-free latch implementation for shared memory."""

import struct
import time
import threading
from typing import Optional

from .memory import Memory


class Latch:
    """
    Lock-free latch for one-time countdown synchronization.

    A latch counts down from an initial value to zero. Once it reaches zero,
    it stays at zero (one-time use). Any number of processes can wait() for
    the count to reach zero, and any process can count_down() to decrement it.

    Unlike Barrier which resets and cycles through generations, Latch is
    single-use and cannot be reset.

    Thread-safe and process-safe.

    Common use cases:
    - Start gate: Initialize with count=1, workers wait(), coordinator counts down
    - Completion detection: Initialize with count=N, each worker counts down when done

    Example:
        # Wait for all workers to initialize
        ready_latch = Latch(mem, "workers_ready", count=num_workers)

        # Each worker thread:
        initialize()
        ready_latch.count_down()  # Signal ready

        # Coordinator:
        ready_latch.wait()  # Wait for all workers to be ready
        start_work()
    """

    HEADER_FORMAT = 'iiii'  # count, initial_count, padding[2]
    HEADER_SIZE = struct.calcsize(HEADER_FORMAT)

    def __init__(self, memory: Memory, name: str,
                 count: Optional[int] = None):
        """
        Create or open a latch.

        Args:
            memory: Memory instance
            name: Latch identifier
            count: Initial count value (required for creation, None to open)

        Raises:
            ValueError: If parameters are invalid
            RuntimeError: If latch not found when opening
        """
        self.memory = memory
        self.name = name

        # Try to find existing latch
        entry = memory.table.find(name)

        if entry is None:
            # Create new latch
            if count is None:
                raise RuntimeError(f"Latch not found: {name}")

            if count < 0:
                raise ValueError("Count must be non-negative")

            # Allocate in shared memory
            self.offset = memory.allocate(name, self.HEADER_SIZE)

            # Initialize header
            header_data = struct.pack(
                self.HEADER_FORMAT,
                count,          # count
                count,          # initial_count
                0,              # padding
                0               # padding
            )
            memory.data[self.offset:self.offset + self.HEADER_SIZE] = header_data

            self.initial_count_value = count

        else:
            # Open existing latch
            self.offset = entry.offset

            if entry.size != self.HEADER_SIZE:
                raise RuntimeError(f"Invalid latch size: {entry.size}")

            # Read initial_count from header
            _, self.initial_count_value, _, _ = self._read_header()

        # Create atomic views for signed integers
        self._count_offset = self.offset
        self._lock = threading.Lock()  # For atomic compare-exchange

    def _read_header(self):
        """Read header values."""
        header_bytes = self.memory.data[self.offset:self.offset + self.HEADER_SIZE]
        return struct.unpack(self.HEADER_FORMAT, header_bytes)

    def _load_count(self) -> int:
        """Atomically load count value."""
        return struct.unpack_from('<i', self.memory.data, self._count_offset)[0]

    def _store_count(self, value: int):
        """Atomically store count value."""
        struct.pack_into('<i', self.memory.data, self._count_offset, value)

    def _compare_exchange_count(self, expected: int, desired: int) -> bool:
        """Atomically compare and exchange count value."""
        with self._lock:
            current = self._load_count()
            if current == expected:
                self._store_count(desired)
                return True
            return False

    def count_down(self, n: int = 1):
        """
        Decrement the count by n (default 1).

        Atomically decrements the count, saturating at 0. If the count
        reaches 0, all waiting processes are released.

        Args:
            n: Amount to decrement (default 1, must be > 0)

        Raises:
            ValueError: If n <= 0
        """
        if n <= 0:
            raise ValueError("count_down amount must be positive")

        current = self._load_count()

        while current > 0:
            new_count = max(0, current - n)

            if self._compare_exchange_count(current, new_count):
                return

            # CAS failed, current was updated, retry
            current = self._load_count()

    def wait(self, timeout: Optional[float] = None) -> bool:
        """
        Wait for the count to reach zero.

        Blocks until the latch count reaches 0. If the count is already 0,
        returns immediately.

        Uses spin-waiting with exponential backoff to reduce CPU usage.

        Args:
            timeout: Maximum time to wait in seconds (None for infinite)

        Returns:
            True if count reached 0, False if timed out
        """
        start_time = time.time() if timeout is not None else None

        backoff = 0.0001  # 0.1ms
        max_backoff = 0.001  # 1ms

        while self._load_count() > 0:
            # Check timeout
            if timeout is not None:
                elapsed = time.time() - start_time
                if elapsed >= timeout:
                    return False

            # Exponential backoff to reduce CPU usage
            time.sleep(backoff)
            if backoff < max_backoff:
                backoff *= 2

        return True

    def try_wait(self) -> bool:
        """
        Try to wait without blocking.

        Returns:
            True if count is 0, False if still counting down
        """
        return self._load_count() == 0

    @property
    def count(self) -> int:
        """Get current count value."""
        return self._load_count()

    @property
    def initial_count(self) -> int:
        """Get initial count value that the latch was created with."""
        return self.initial_count_value

    def __repr__(self):
        return (f"Latch(name='{self.name}', count={self.count}, "
                f"initial_count={self.initial_count})")

File: go/zeroipc/latch.go
package zeroipc

import (
	"encoding/binary"
	"errors"
	"fmt"
	"sync/atomic"
	"time"
	"unsafe"
)

// LatchHeaderSize is the size of the latch header in bytes.
// Layout: count(4) + initial_count(4) + padding(8) = 16 bytes
const LatchHeaderSize = 16

// Latch is a lock-free one-time countdown synchronization primitive.
//
// A latch counts down from an initial value to zero. Once it reaches zero,
// it stays at zero (one-time use). Any number of processes can Wait() for
// the count to reach zero, and any process can CountDown() to decrement it.
//
// Unlike Barrier which resets and cycles, Latch is single-use.
//
// Binary layout:
//   - count: int32 (atomic, offset 0) - current count
//   - initial_count: int32 (offset 4) - initial count (immutable)
//   - _padding: [2]int32 (offset 8) - alignment padding
//
// Common use cases:
//   - Start gate: count=1, workers Wait(), coordinator CountDown()
//   - Completion: count=N, each worker CountDown(), coordinator Wait()
type Latch struct {
	memory       *Memory
	name         string
	offset       int
	initialCount int32
}

// NewLatch creates a new latch in shared memory.
func NewLatch(memory *Memory, name string, count int32) (*Latch, error) {
	if len(name) >= NameSize {
		return nil, errors.New("name too long (max 31 characters)")
	}

	if count < 0 {
		return nil, errors.New("count must be non-negative")
	}

	// Check if already exists
	if entry := memory.Find(name); entry != nil {
		return nil, fmt.Errorf("latch '%s' already exists", name)
	}

	// Allocate space
	offset, err := memory.Allocate(name, LatchHeaderSize)
	if err != nil {
		return nil, fmt.Errorf("failed to allocate: %w", err)
	}

	data := memory.Data()

	// Write header
	binary.LittleEndian.PutUint32(data[offset:], uint32(count))   // count
	binary.LittleEndian.PutUint32(data[offset+4:], uint32(count)) // initial_count
	binary.LittleEndian.PutUint32(data[offset+8:], 0)             // padding[0]
	binary.LittleEndian.PutUint32(data[offset+12:], 0)            // padding[1]

	return &Latch{
		memory:       memory,
		name:         name,
		offset:       offset,
		initialCount: count,
	}, nil
}

// OpenLatch opens an existing latch in shared memory.
func OpenLatch(memory *Memory, name string) (*Latch, error) {
	entry := memory.Find(name)
	if entry == nil {
		return nil, fmt.Errorf("latch '%s' not found", name)
	}

	if entry.Size != LatchHeaderSize {
		return nil, fmt.Errorf("invalid latch size: %d, expected %d", entry.Size, LatchHeaderSize)
	}

	offset := int(entry.Offset)
	data := memory.Data()

	initialCount := int32(binary.LittleEndian.Uint32(data[offset+4:]))

	return &Latch{
		memory:       memory,
		name:         name,
		offset:       offset,
		initialCount: initialCount,
	}, nil
}

// countPtr returns a pointer to the atomic count.
func (l *Latch) countPtr() *int32 {
	return (*int32)(unsafe.Pointer(&l.memory.Data()[l.offset]))
}

// CountDown decrements the count by n (default 1).
// Saturates at 0 (never goes negative).
func (l *Latch) CountDown(n int32) error {
	if n <= 0 {
		return errors.New("count_down amount must be positive")
	}

	for {
		current := atomic.LoadInt32(l.countPtr())

		if current <= 0 {
			return nil // Already at 0
		}

		newCount := current - n
		if newCount < 0 {
			newCount = 0
		}

		if atomic.CompareAndSwapInt32(l.countPtr(), current, newCount) {
			return nil
		}
		// CAS failed, retry
	}
}

// CountDownOne decrements the count by 1.
func (l *Latch) CountDownOne() {
	_ = l.CountDown(1)
}

// Wait waits for the count to reach zero.
// If already at 0, returns immediately.
func (l *Latch) Wait() {
	backoff := time.Microsecond
	maxBackoff := time.Millisecond

	for atomic.LoadInt32(l.countPtr()) > 0 {
		time.Sleep(backoff)
		if backoff < maxBackoff {
			backoff *= 2
		}
	}
}

// TryWait checks if the count has reached zero without blocking.
func (l *Latch) TryWait() bool {
	return atomic.LoadInt32(l.countPtr()) == 0
}

// WaitTimeout waits for the count to reach zero with a timeout.
// Returns true if count reached 0, false if timed out.
func (l *Latch) WaitTimeout(timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	backoff := time.Microsecond
	maxBackoff := time.Millisecond

	for atomic.LoadInt32(l.countPtr()) > 0 {
		if time.Now().After(deadline) {
			return false
		}

		time.Sleep(backoff)
		if backoff < maxBackoff {
			backoff *= 2
		}
	}

	return true
}

// Count returns the current count value.
func (l *Latch) Count() int32 {
	return atomic.LoadInt32(l.countPtr())
}

// InitialCount returns the initial count value.
func (l *Latch) InitialCount() int32 {
	return l.initialCount
}

// Name returns the latch name.
func (l *Latch) Name() string {
	return l.name
}

// Reset resets the latch to its initial count.
//
// WARNING: This is NOT thread-safe if other goroutines/processes are
// waiting or counting down. Only use with exclusive access.
func (l *Latch) Reset() {
	atomic.StoreInt32(l.countPtr(), l.initialCount)
}
